modules/optimization/src/optimizationreporters/GeneralOptimization.C

```diff
@@ -0,0 +1,114 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MooseError.h"
+#include "MooseTypes.h"
+#include "GeneralOptimization.h"
+#include "libmesh/id_types.h"
+#include <numeric>
+
+registerMooseObject("OptimizationApp", GeneralOptimization);
+
+InputParameters
+GeneralOptimization::validParams()
+{
+  InputParameters params = OptimizationReporterBase::validParams();
+
+  params.addRequiredParam<ReporterValueName>("objective_name", "Objective reporter name.");
+  params.addParam<std::vector<dof_id_type>>(
+      "num_values",
+      "Number of parameter values associated with each parameter group in 'parameter_names'.");
+  params.addParam<ReporterValueName>("num_values_name",
+                                     "Reporter that holds the number of parameter values "
+                                     "associated with each parameter group in 'parameter_names'.");
```

Comment:
I am thinking of two scenarios. 
1: The number of parameters is set regardless of any
changes in the forward problem aka mesh refinement does not affect number of
parameters.
2: The number of parameters changes as the mesh gets refined for example
individual elemental material properties. 
When I say mesh gets refined, I don't mean in the middle of the solve, just a
totally new run. Having option 2 makes it so the user doesn't have to run the
forward problem get the number of parameters of the new mesh then update that in
the optimization app.

Reply:
Would there be a useful error message if the number of parameters didn't match what is expected in the forward problem?

Reply:
Most likely the forward problem would have to output that message. I don't think
I am 100% understanding what you are asking, though. For scenario 2, the forward
problem determines the number of parameters so there shouldn't be a mismatch.
For scenario 1, it would have the same error message that would happen normally
if specifying incorrect number of parameters.

Reply:
I guess I'm fine with this. I hope in the future this can be setup automatically.